python/llm/src/ipex_llm/transformers/npu_models/linear.py

```diff
@@ -125,6 +131,239 @@ def fromTensor(
+    else:
+        invalidInputError(False, f"Unsupported dtype for weights {weights.dtype}")
+
+    # Use or not op_id depending on the class used
+    op_kwargs = {"op_id": op_id} if op_id else {}
+
+    original_input_shape = x.shape
+    expected_output_shape = list(original_input_shape[:-1]) + [outC]
+
+    # Reshape input
+    input_dtype = x.dtype
+    x = x.to(torch.float16) if input_dtype != torch.float16 else x
+    if len(x.shape) > 2 or x.shape[-1] != inC:
+        x = x.view([-1, inC])
+    x_np = x.numpy()
+    batch = x_np.shape[0]
+
+    key = f"{str(op_class_name)}_{batch}_{inC}_x_{outC}_{inC}_{x_np.dtype}"
+    models = _model_cache.get(key, None)
+
+    if models is None:
+        _model_cache[key] = deque([create_op(inC, outC, batch)])
+
+    # Get the model
+    model = _model_cache[key][0]
+
+    ret = model.run(x_np, *op_args, **op_kwargs)
+
+    return adapt_output_tensor(ret, expected_output_shape, input_dtype)
+
+
+def adapt_output_tensor(
+    output: np.ndarray, original_shape: torch.Size, input_dtype: torch.dtype
+) -> torch.Tensor:
+    """Adapt the output tensor to the original shape and dtype.
+
+    Args:
+        output (np.ndarray): output tensor
+        original_shape (torch.Size): original shape
+        input_dtype (torch.dtype): input dtype
+
+    Returns:
+        torch.Tensor: output tensor
+    """
+    output = torch.from_numpy(output)
+    if output.shape != original_shape:
+        output = output.view(original_shape)
+    # needs to copy as the same buffer can be reutilized
+    return output.to(input_dtype, copy=True)
+
+
+def set_contiguous(tensor: torch.Tensor) -> torch.Tensor:
+    """Set tensor to be contiguous in memory.
+
+    Args:
+        tensor (torch.Tensor): input tensor
+
+    Returns:
+        torch.Tensor: output, contiguous tensor
+    """
+    if not tensor.is_contiguous():
+        return tensor.contiguous()
+    return tensor
+
+
+class LMHeadLinear(NNFactory):
+    """Quantized Linear class, computing a matrix matrix multiplication with weights prefetching."""
+
+    def __init__(
+        self,
+        inC: int,
+        outC: int,
+        batch: int,
+        split_num: int = 2,
+        profile: bool = False,
+        device: str = "NPU",
+        dtype: np.dtype = np.int8,
+    ):
+        """Initialize the QLinear class.
+
+        Args:
+            inC (int): input channels
+            outC (int): output channels
+            batch (int): batch
+            profile (bool): Enable/Disable profiling. Defaults to False.
+            device (str): Target device, default to "NPU".
+            dtype (np.dtype): weights datatype. Defaults to np.int8.
+
+        """
+        super().__init__(profile, device)
+        self.inC, self.outC = inC, outC
+        self.batch = batch
+
+        input = self.parameter((self.batch, self.inC))
+
+        self.split_num = split_num
+        split_size = self.inC // split_num // 2 * 2
+
+        for i in range(7):
```

Comment:
why 7?

Reply:
as we use 7 as split_num for qwen2's lm_head